src/autoproj.ts

```diff
@@ -166,6 +235,158 @@ export class Workspace
             })
         })
     }
+
+    private runCommandToCompletion(process, error?: string) : Promise<void> {
+        return new Promise<void>((resolve, reject) => {
+            process.on('exit', (code, signal) => {
+                if (code == 0) {
+                    resolve();
+                }
+                else {
+                    reject(new Error(error));
+                }
+            });
+        });
+    }
+
+    public syskitDefaultBundle() : string {
+        return path.join(this.root, '.vscode', 'rock-default-bundle');
+    }
+
+    syskitGenApp(path: string) : Promise<void> {
+        let process = this.autoprojExec("syskit", ["gen", "app", path]);
+        this.redirectProcessToChannel(`syskit gen ${path}`, "gen", process);
+        return this.runCommandToCompletion(process, `failed to run \`syskit gen app ${path}\``);
+    }
+
+    async syskitCheckApp(path: string) : Promise<void> {
+        let process = this.autoprojExec("syskit", ["check", path]);
+        this.redirectProcessToChannel(`syskit check ${path}`, "check", process);
+        return this.runCommandToCompletion(process, `bundle in ${path} seem invalid, or syskit cannot be executed in this workspace`);
+    }
+
+    public defaultBundlePath() : string {
+        return path.join(this.root, '.vscode', 'rock-default-bundle');
+    }
+
+    public hasValidSyskitContext() : Promise<boolean> {
+        // We do the cheap existence check even if the syskit context has been
+        // verified. This would allow the user to "reset" the bundle by deleting
+        // the bundle folder without having to restart VSCode, with a small
+        // performance cost
+        let bundlePath = this.defaultBundlePath();
+        if (!fs.existsSync(bundlePath)) {
+            return Promise.resolve(false);
+        }
+
+        if (this._verifiedSyskitContext) {
+            return Promise.resolve(true);
+        }
+
+        return this.syskitCheckApp(bundlePath).
+            then(() => {
+                this._verifiedSyskitContext = true;
+                return true;
+            }).
+            catch(() => false);
+    }
+
+    public ensureSyskitContextAvailable(): Promise<void>
+    {
+        let pending = this._pendingWorkspaceInit;
+        if (pending) {
+            return pending;
+        }
+
+        let p = this.hasValidSyskitContext().then((result) => {
+            if (result) {
+                this._pendingWorkspaceInit = undefined;
+            }
+            else {
+                let bundlePath = this.defaultBundlePath();
+                return this.syskitGenApp(bundlePath).
+                    then(
+                        ()  => { this._pendingWorkspaceInit = undefined },
+                        (e) => { this._pendingWorkspaceInit = undefined; throw e; }
+                    );
+            }
+        })
+
+        this._pendingWorkspaceInit = p;
+        return p;
+    }
+
+    private redirectProcessToChannel(name, shortname, process)
```

Comment:
> Typescript's strong typing is a cool feature, I don't understand why you avoid using it..

Given the amount of time I've spent fighting it, especially when mocking, I wouldn't say that "I'm not using it". Rather than I'm using it less religiously than you.

The tests do spot the problem, actually ... just didn't run them (as I noticed above, I still have to write tests for a lot of the new stuff).

Reply:
> The tests do spot the problem, actually

Actually not, interestingly enough, since node.js' `process` is has `stdout` and `stderr` and both of them are event emitters ... A.k.a. "same duck type". One point for you ;-)